File: Part 2 - Regression/Section 6 - Polynomial_Regression/wil_polynomial_regression.R
# Polynomial regression
library(caTools)
library(ggplot2)


# Variables
data.directory <- "."
data.file <- "Position_Salaries.csv"

# Load the data
data.set <- read.csv(file.path(data.directory, data.file))
data.set <- data.set[2:3]
head(data.set)

# Fitting linear model first
lin.reg <- lm(data = data.set, formula = Salary ~ .)

# Poly regression
data.set$Level2 <- data.set$Level^2
data.set$Level3 <- data.set$Level^3
data.set$Level4 <- data.set$Level^4
poly.reg <- lm(formula = Salary ~ ., data = data.set)
summary(poly.reg)

# Visualise the linear model
ggplot() +
  geom_point(aes(x = data.set$Level, y = data.set$Salary), color = 'red') +
  geom_line(aes(x = data.set$Level, y = predict(lin.reg, newdata = data.set) ), color = 'blue') +
  ggtitle('The title') +
  xlab('Level') +
  ylab('Salary')


# Visualise the linear model
x_grid = seq(min(data.set$Level), max(data.set$Level), 0.1)

ggplot() +
  geom_point(aes(x = data.set$Level, y = data.set$Salary), color = 'red') +
  geom_line(aes(x = x_grid, y = predict(poly.reg, newdata = data.frame(Level = x_grid,
                                                                       Level2 = x_grid^2,
                                                                       Level3 = x_grid^3,
                                                                       Level4 = x_grid^4)) ), color = 'blue') +
  ggtitle('The title') +
  xlab('Level') +
  ylab('Salary')


# Predict the value of 6.5 level
lin.pred <- predict(lin.reg,newdata = data.frame(Level = 6.5))

# Use Polynomial model
poly.pred <- predict(poly.reg,newdata = data.frame(Level = 6.5,
                                                   Level2 = 6.5^2,
                                                   Level3 = 6.5^3,
                                                   Level4 = 6.5^4))

File: Part 2 - Regression/Section 6 - Polynomial_Regression/wil_polynomial_regression.py
#!/usr/bin/env python

# File information
"""
wil_polynomial_regression.py: First look at polynomial regression of data.
"""
__author__ = "Wil Poole"
__copyright__ = "2017"
__credits__ = "Wil Poole"
__license__ = "None"
__version__ = "0.1"
__maintainer__ = "Wil Poole"
__email__ = "@"
__status__ = "Development"


# Import libraries
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns


# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
print(dataset.head)
# Upper bound of range is always ignored, but need as a matrix, so include 1:2
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

plot = plt.scatter(X, y, color = "red")
plt.show(plot)

# Fit the linear model
from sklearn.linear_model import LinearRegression
linreg = LinearRegression()
linreg.fit(X, y)

print(linreg)

# Polyomial
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)

# Fit the model
linreg2 = LinearRegression()
linreg2.fit(X_poly,y)

# plot
plt.scatter(X,y,color = 'red')
plt.plot(X,linreg.predict(X), color = 'blue')
plt.title('What')
plt.xlabel('Real')
plt.ylabel('Model')
plt.show()

# plot the polynomial
X_grid = np.arange(min(X),max(X),0.1) # For a smoother curve
X_grid = X_grid.reshape(len(X_grid),1)
plt.scatter(X,y,color = 'red')
plt.plot(X_grid,linreg2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('What')
plt.xlabel('Real')
plt.ylabel('Model')
plt.show()

# Predict using the linear model
linreg.predict(6.5)

# Predict using the polynomial
linreg2.predict(poly_reg.fit_transform(6.5))
